File: solutions/rust/forth/1/src/handlers/default_handlers.rs
use std::collections::HashMap;

use crate::Error;
use crate::handlers::handler_function::FunctionHandler;
use crate::handlers::math_helpers::{*};
use crate::handlers::stack_helpers::{*};

static mut DEFAULT_HANDLERS: Option<HashMap<&'static str, FunctionHandler>> = None;

pub fn get_default_handlers() -> Result<&'static HashMap<&'static str, FunctionHandler>, Error> {
	unsafe {
		if DEFAULT_HANDLERS.is_none() {
			DEFAULT_HANDLERS = Some(initialize_default_handlers());
		}

		match DEFAULT_HANDLERS.as_ref() {
			Some(handler) => Ok(handler),
			None => Err(Error::InvalidWord)
		}
	}
}

fn initialize_default_handlers() -> HashMap<&'static str, FunctionHandler> {
	return HashMap::from([
		("+", FunctionHandler::Handle(add)),
		("-", FunctionHandler::Handle(subtract)),
		("*", FunctionHandler::Handle(multiply)),
		("/", FunctionHandler::Handle(divide)),
		("dup", FunctionHandler::Handle(duplicate_last)),
		("drop", FunctionHandler::Handle(drop_last)),
		("swap", FunctionHandler::Handle(swap_last)),
		("over", FunctionHandler::Handle(penultimate_value_copy)),
	]);
}

File: solutions/ocaml/forth/8/default_handlers.ml
type t = (string, Context.t -> Context.t) Hashtbl.t

let create () : t = Hashtbl.create 16

let add (handlers : t) key ~(f : Context.t -> Context.t) : unit =
  let lowercase_key = String.lowercase_ascii key in
  Hashtbl.replace handlers lowercase_key f

let get (handlers : t) (key : string) : (Context.t -> Context.t) option =
  let lowercase_key = String.lowercase_ascii key in
  Hashtbl.find_opt handlers lowercase_key

let create_default () : t =
  let handlers = create () in
  add handlers "+" ~f:Math_helpers.add;
  add handlers "-" ~f:Math_helpers.subtract;
  add handlers "*" ~f:Math_helpers.multiply;
  add handlers "/" ~f:Math_helpers.divide;
  add handlers "dup" ~f:Stack_helpers.duplicate_last;
  add handlers "drop" ~f:Stack_helpers.drop_last;
  add handlers "swap" ~f:Stack_helpers.swap_last;
  add handlers "over" ~f:Stack_helpers.penultimate_value_copy;
  handlers
